File: src/CrystalMarkNet/EntityFX.CrystalMarkNet/ALU/Eratosthenes.cs
using System.Threading;

namespace EntityFX.CrystalMarkNet.ALU
{
    class Eratosthenes : CrystalBenchmarkBase
    {
        private const int N = 1024;

        protected override int BenchImplementation(CancellationToken cancellationToken)
        {
            int i, p, k, count = 0, cnt;
            bool[] flag = new bool[N + 1];

            for (; ; )
            {
                cnt = 1;
                for (i = 0; i < N; i += 4)
                {
                    flag[i] = true;
                    flag[i + 1] = true; 
                    flag[i + 2] = true; 
                    flag[i + 3] = true;
                }
                flag[N] = true;

                for (i = 0; i <= N; i++)
                {
                    if (flag[i])
                    {
                        p = i + i + 3;
                        for (k = i + p; k <= N; k += p)
                        {
                            flag[k] = false;
                        }
                        cnt++;
                    }
                }
                count++;

                if (cancellationToken.IsCancellationRequested)
                {
                    return (int)(count * (N / 70000.0 / 10));
                }
            }
        }
    }
}

File: src/lua/alu/eratosthenes.lua
Eratosthenes = class(CrystalBenchmarkBase, function(eratosthenes, writer)
    CrystalBenchmarkBase.init(eratosthenes, writer, "Eratosthenes")
end)

Eratosthenes.N = 1024

function Eratosthenes:benchImplementation()
    local p, k, count, cnt = 0, 0, 0, 0
    local flag = {}

    while not self:isCancelled() do
        cnt = 1

        for i = 1, Eratosthenes.N, 1 do flag[i] = true end

        for i = 1, Eratosthenes.N, 1 do
            if flag[i] then
                p = i + i + 3
                k = i + p
                while k <= Eratosthenes.N do
                    flag[k] = false
                    k = k + p
                end
                cnt = cnt + 1
            end

        end

        count = count + 1
    end

    return count / 1000.0 / 100.0 * Eratosthenes.N
end
